managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -268,15 +277,53 @@ protected void recoverFromLedger(final ManagedCursorInfo info, final VoidCallbac
                 }
 
                 PositionImpl position = new PositionImpl(positionInfo);
+                PositionImpl lastIndividualAckedMessage = null;
                 if (positionInfo.getIndividualDeletedMessagesCount() > 0) {
                     recoverIndividualDeletedMessages(positionInfo.getIndividualDeletedMessagesList());
+                    // also recover readPosition to last acked message
+                    lastIndividualAckedMessage = new PositionImpl(
+                            positionInfo.getIndividualDeletedMessagesList().get(positionInfo.getIndividualDeletedMessagesCount() - 1).getUpperEndpoint());
                 }
                 recoveredCursor(position);
+                // if individualDeletedMessages are recovered then readPosition should be set after last Acked-Message
+                if (lastIndividualAckedMessage != null) {
+                    this.readPosition = lastIndividualAckedMessage;
+                }
                 callback.operationComplete();
             }, null);
         }, null);
     }
 
+    /**
+     * return unacked messages which are not part of individual deleted messages
+     * 
+     * @return
+     */
+    @Override
+    public Set<? extends Position> getNotDeletedMessages() {
```

Comment:
OK. In this case then, we should skip the read position to the next available message. 

Eg. : 
 * Mark-delete: `1:0`, 
 * Individually deleted messages: `[(1:1..1:10]]` 

In this case `1:1` was not acked.

If I try to read 5 messages, I should get a list with `1:1` and the `readPosition` should be moved to `1:11`.

This could be implemented such that, if I read and some of the messages were filtered, we check for the next unacked message to move the read position for next time.

My point here is that this issue should be handled at the ManagedLedger layer. By returning the unacked messages to the application we're kind of leaking back the logic to the broker.
Also, it might not only happen for the shared subscription.